llm/gemma/scripts/lora.py

```diff
@@ -0,0 +1,152 @@
+    checkpoints = [c for c in checkpoints if c.name.split('-')[-1].isdigit()]
+    checkpoints = sorted(checkpoints,
+                         key=lambda x: int(x.name.split('-')[-1]),
+                         reverse=True)
+    for checkpoint in checkpoints:
+        if not (checkpoint / 'complete').exists():
+            print(f'Removing incomplete checkpoint {checkpoint}')
+            shutil.rmtree(checkpoint)
+        else:
+            print(f'Using checkpoint {checkpoint}, copying to ~/tmp/ for '
+                  'optimization of loading.')
+            tmp_dir = os.path.expanduser('~/tmp')
+            os.makedirs(tmp_dir, exist_ok=True)
+            try:
+                # Optimization for checkpoint loading. This is to force the
+                # mounting tool to download the checkpoints in parallel first.
+                # It will improve the loading speed of the checkpoints
+                # significantly.
+                subprocess.run(
+                    ['gsutil', '-m', 'rsync', '-r', checkpoint, tmp_dir],
+                    check=True)
+            except:
+                print('Failed to optimize checkpoint loading. Skip.')
+            break
+
+
+def train():
+    parser = transformers.HfArgumentParser((ModelArguments, TrainingArguments))
+    model_args, training_args = parser.parse_args_into_dataclasses()
+    local_rank = training_args.local_rank
+    if local_rank == 0:
+        cleanup_incomplete_checkpoints(training_args.output_dir)
+    torch.distributed.barrier()
+
+    # Check the existence of checkpoints in all processes
+    # All ranks must simultaneously resume from a checkpoint if it exists.
+    # Otherwise, upon recovery the model weights may not reload correctly,
+    # causing loss spikes.
+    resume_from_checkpoint = False
+    checkpoints = list(
+        pathlib.Path(training_args.output_dir).glob('checkpoint-*'))
+    checkpoints = [c for c in checkpoints if c.name.split('-')[-1].isdigit()]
+    if checkpoints:
+        resume_from_checkpoint = True
+
+    bnb_config = BitsAndBytesConfig(load_in_4bit=True,
+                                    bnb_4bit_quant_type="nf4",
+                                    bnb_4bit_compute_dtype=torch.bfloat16)
+
+    tokenizer = AutoTokenizer.from_pretrained(model_args.model_name_or_path,
+                                              token=os.environ['HF_TOKEN'])
+    model = AutoModelForCausalLM.from_pretrained(model_args.model_name_or_path,
+                                                 quantization_config=bnb_config,
+                                                 device_map="auto",
+                                                 token=os.environ['HF_TOKEN'])
+
+    lora_config = LoraConfig(
+        r=8,
+        target_modules=[
+            "q_proj", "o_proj", "k_proj", "v_proj", "gate_proj", "up_proj",
+            "down_proj"
+        ],
+        task_type="CAUSAL_LM",
+    )
+
+    data = load_dataset("Abirate/english_quotes")
```

Comment:
Shall we make this an env as well? Are there >1 common datasets we can list in the README?